website/www/site/content/en/blog/apache-beam-flink-and-kubernetes.md

```diff
@@ -0,0 +1,355 @@
+
+
+Alerting is threshold based with frequent metrics checking. If a threshold is met, and continues for a certain amount of times, we will alert our internal Slack channels, or PagerDuty for immediate attention. We tune the alerting so that the accuracy is high.
+
+
+# Cost Optimization: Strategies and Tuning
+
+A major goal for us to move to a self managed streaming service is cost efficiency. We have achieved the goal by saving more than half of the cost by several minor tunings. We can see more room for improvements further.
+
+These are just a few tips in our journey so far:
+
+
+
+1. Using GCS as Checkpointing storage.
+2. Reducing the write frequency to GCS
+3. Freedom to use different machine types. For example, in GCP N2D machines are 15% cheaper than N2 machines.
+4. Autoscaling tasks to use optimal resources while maintaining latency SLA (we will cover the details in Part 2)
+
+We describe more details for #1 and #2 related to checkpointing optimization as follows.
+
+**Google Cloud Storage and Checkpointing**
+
+GCS serves as our checkpoint store due to its cost-effectiveness, scalability, and durability. When working with GCS, several design considerations and best practices can optimize scaling and performance:
+
+
+
+* Data partitioning methods like range partitioning, which divides data based on specific attributes, and hash partitioning, distributing data evenly using hash functions, are crucial.
+* Avoiding sequential key names, especially timestamps, helps prevent hotspots and uneven data distribution; instead, introduce random prefixes for object distribution.
+* Employing a hierarchical folder structure enhances data management and reduces the number of objects in a single directory.
+* For small files, combining them into larger ones improves read throughput, while minimizing small files reduces inefficient storage use and metadata operations.
+
+**Tune the frequency writing the GCS**
+
+One of our primary challenges lay in scaling jobs efficiently. Stateless jobs, which are relatively simpler, still presented hurdles, especially in scenarios where Flink needed to process an overwhelming number of workers. To overcome this challenge, We increased state.storage.fs.memory-threshold settings to 1 MB from 20KB (??). This configuration allowed us to combine small checkpoints files into larger ones at Job Manager level and reduce metadata calls.
+
+Another challenge we tackled was optimizing the performance of GCS operations. While GCS is excellent for streaming large amounts of data, it does have limitations when it comes to handling high-frequency I/O requests. To mitigate this, we employed strategies like introducing random prefixes in key names, avoiding sequential key names, and optimizing our GCS sharding techniques. These methods significantly enhanced our GCS performance, enabling smooth operation of our stateless jobs.
+
+This is the chart showing the GCS writes reduction after changing the memory-threshold:
```

Comment:
```suggestion
The following chart shows the Google Cloud Storage writes reduction after changing the memory-threshold:
```

Reply:
Resolving this as it is flagged outdated.